### ISSUE TITLE
UTF-8 BOM is not removed when running on Node.js and input is a file

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor.js#1344]
Although there's a test that covers removing the UTF-8 BOM from the AsciiDoc source, this only works on Node.js when the input is a string created in JavaScript. When the AsciiDoc source is read from a file, or the string is created from Buffer.from, the UTF-8 BOM is not removed.

This snippet explains why we're hitting this problem:

```js
const fs = require('fs')

fs.writeFileSync('test.adoc', Buffer.concat([Buffer.from([0xEF, 0xBB, 0xBF]), Buffer.from('= Document Title')]))

const contentsFromFile = fs.readFileSync('test.adoc')
console.log(contentsFromFile.toString().charCodeAt())
console.log(contentsFromFile.toString().charCodeAt(1))
// => 65279
// => '='

const contentsFromString = '\xef\xbb\xbf= Document Title'
console.log(contentsFromString.charCodeAt())
console.log(contentsFromString.charCodeAt(3))
// => 239
// => '='
```

It appears that when Node.js creates a string by way of a Buffer, such as when reading the contents of a file, it changes the UTF-8 BOM into a different BOM character (code: 65279, char ref: 0xFEFF). I have not found any way to disable this behavior. It's basically a quirk of Node.js.

I think Asciidoctor.js should detect this alternate BOM and remove the character. (I'm open to changing Asciidoctor Ruby, if we determine it's necessary).

**Comment by mojavelinux** [comment]
What I can't figure out is how to emulate this scenario in Ruby. I can't figure how to get Ruby to report the character code (8-bit unsigned integer) 65279. If I try to pack it, this is what I get:

```rb
[65279].pack('C*')
# => "\xFF"
```

Here's what I get if I unpack the BOM:

```rb
"\xFE\xFF".unpack('C*')
=> [254, 255]
```

We may just have to accept this character code is a weird quirk of Node.js and deal with it as such.

**Comment by mojavelinux** [comment]
I'm now convinced it's best to deal with this in Asciidoctor.js. Our custom unpack method is an ideal place for it. Here's the logic we could use that would make this work:

```
`self.charCodeAt() === 65279 ? [239, 187, 191] : #{self[0, 3].bytes.select.with_index {|_, i| i.even? }}`
```

Let me know if you want me to proceed with a fix.

**Comment by mojavelinux** [comment]
An alternate way to do this to check if `self[0, 1].bytes` is `[255, 254]`. Either way seems to work.

**Comment by mojavelinux** [comment]
Here's something that's interesting:

```js
encodeURIComponent(self.charAt())
// => %EF%BB%BF
```

and

```js
encodeURIComponent(String.fromCharCode(65279))
// => %EF%BB%BF
```

So it **is** recognizing 65279 as a UTF-8 BOM.

I think checking the char code is the right strategy. It's the same as:

```js
self.charCodeAt() === String.fromCharCode(65279)
```

**Comment by mojavelinux** [comment]
Now I understand what you were trying to point out. You are correct, we need to also patch String#byteslice to cover the same scenario. I had confused myself since I was using a modified version of Asciidoctor core. When I cleaned everything and started over, things started working.

As to why I was having difficulty getting tests to use the updated code, I needed to discover `npm run clean:core` and `npm run clean:patch`. Very useful.
